snapcraft/plugins/v2/colcon.py

```diff
@@ -0,0 +1,230 @@
+                    "type": "array",
+                    "minitems": 1,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "colcon-packages-ignore": {
+                    "type": "array",
+                    "minitems": 1,
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+            },
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {
+            "python3-colcon-common-extensions",
+            "python3-rosdep",
+            "python3-wstool",
+            "python3-rosinstall",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "AMENT_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "COLCON_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "ROS_PYTHON_VERSION": "3",
+        }
+
+    def _get_colcon_build_command(self) -> str:
+        cmd: List[str] = [
+            "colcon",
+            "build",
+            "--merge-install",
+            "--install-base",
+            "$SNAPCRAFT_PART_INSTALL",
+        ]
+
+        if self.options.colcon_packages_ignore:
+            cmd.extend(["--packages-ignore", *self.options.colcon_packages_ignore])
+
+        if self.options.colcon_packages:
+            cmd.extend(["--packages-select", *self.options.colcon_packages])
+
+        if self.options.colcon_ament_cmake_args:
+            cmd.extend(["--ament-cmake-args", *self.options.colcon_ament_cmake_args])
+
+        if self.options.colcon_catkin_cmake_args:
+            cmd.extend(["--catkin-cmake-args", *self.options.colcon_catkin_cmake_args])
+
+        # Specify the number of workers
+        cmd.extend(["--parallel-workers", "${SNAPCRAFT_PARALLEL_BUILD_COUNT}"])
+
+        return " ".join(cmd)
+
+    def _get_stage_runtime_dependencies_command(self):
+        env = dict(
+            (key, os.environ[key])
+            for key in ["SNAP", "SNAP_ARCH", "SNAP_NAME", "SNAP_VERSION"]
+            if key in os.environ
+        )
+        env["LC_ALL"] = "C.UTF-8"
+        env["LANG"] = "C.UTF-8"
+        env["PYTHONPATH"] = ":".join(["$PYTHONPATH"] + [p for p in sys.path if p])
+
+        env_flags = [f"{key}={value}" for key, value in env.items()]
+
+        return " ".join(
+            [
+                "env",
+                *env_flags,
+                sys.executable,
+                os.path.abspath(__file__),
+                "stage-runtime-dependencies",
+            ]
+        )
+
+    def get_build_commands(self) -> List[str]:
+        return [
+            ". /opt/ros/$ROS_DISTRO/setup.sh",
+            "if [ ! -f /etc/ros/rosdep/sources.list.d/20-default.list ]; then sudo rosdep init; fi",
+            "rosdep update --rosdistro $ROS_DISTRO",
+            "rosdep install --from-paths . -y -r",
```

Comment:
I question the inclusion of `-r`, to continue installing despite errors. What is the reasoning here?

I also suggest including `-i` so we don't install debs that are actually included in the code being built here.

Reply:
Ah, I think I thought it was recursive, removing...

Reply:
Hrm, I may have had issues with -a, I'll bring them back.

Reply:
That's right, can't use `-a` with `--from-paths`.  But they seem to be otherwise equal in this case?  Added `-i` back.

Reply:
Yeah that should work, nice job.